### PR TITLE
Add Piro transient optimization

### PR DESCRIPTION
**Pull request by kliegeois**
@mperego 
@trilinos/rol 

## Motivation

This PR adds Piro capabilities to solve transient PDE constrained optimization problems.
This feature will be used by Albany.

## Testing

This PR adds transient tests with one and two parameters and has been tested in release and debug.
It has been tested with ROL and Teko, without ROL, and with ROL and without Teko.
Albany built and run successfully against that branch.

### CODE REVIEW

File: packages/piro/src/Piro_PerformAnalysis.hpp [hunk 1]
```diff
@@ -78,6 +78,26 @@ namespace Piro {
      );
   //@}
 
+  //! \brief Performs analysis of a solved model using ROL.
+  //! \details Requires that the ROL package is available.
+  //! \ingroup Piro_Thyra_analysis_driver_grp
+  int PerformSROLAnalysis(
```
Comment (mperego) [inline packages/piro/src/Piro_PerformAnalysis.hpp:84]: I would call these: PerformROLSteadyAnalysis and PerformROLTransientAnalysis. Also the descriptions of the two functions need to be updated.

File: packages/piro/src/Piro_TempusReducedObjective.hpp [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+#include "Thyra_VectorStdOps.hpp"
+
+#include "ROL_Objective.hpp"
+#include "ROL_DynamicObjective.hpp"
+#include "ROL_Vector.hpp"
+#include "ROL_ThyraVector.hpp"
+#include "Piro_ROL_ObserverBase.hpp"
+#include "Piro_TempusIntegrator.hpp" 
+
+#include "ROL_ReducedDynamicStationaryControlsObjective.hpp"
```
Comment (mperego) [inline packages/piro/src/Piro_TempusReducedObjective.hpp:68]: Are these three includes needed?

File: packages/piro/src/Piro_TempusReducedObjective.hpp [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+#include "ROL_SerialStationaryControlsConstraint.hpp"
+#include "ROL_SerialStationaryControlsObjective.hpp"
+
+namespace Piro {
+
+template <typename Real>
+class ThyraProductME_TempusFinalObjective : public virtual ROL::DynamicObjective<Real> {
```
Comment (mperego) [inline packages/piro/src/Piro_TempusReducedObjective.hpp:75]: Why is the class name different than the file name?

File: packages/piro/src/Piro_PerformAnalysis.cpp [hunk 4]
```diff
@@ -748,6 +758,340 @@ if(useCustomDotProduct) {
...
+  RolOutputBuffer<char> rolOutputBuffer;
+  std::ostream rolOutputStream(&rolOutputBuffer);
+  Teuchos::RCP<Teuchos::FancyOStream> rolOutput = Teuchos::getFancyOStream(Teuchos::rcpFromRef(rolOutputStream));
+  rolOutput->setOutputToRootOnly(0);
+
+  Teuchos::RCP<Thyra::VectorBase<double> > scaling_vector_p = p->clone_v();
+  ::Thyra::put_scalar<double>( 1.0, scaling_vector_p.ptr());
+  ROL::PrimalScaledThyraVector<double> rol_p_primal(p, scaling_vector_p);
+
+  if ( useFullSpace ) {
```
Comment (mperego) [inline packages/piro/src/Piro_PerformAnalysis.cpp:946]: If the fullSpace code is not tested or if it is not working then we should error out when full space is selected.

File: packages/piro/src/Piro_PerformAnalysis.cpp [hunk 5]
```diff
@@ -748,6 +758,340 @@ if(useCustomDotProduct) {
...
+      return_status = algo->getState()->statusFlag;
+    }  else {
+      *out << "Piro::PerformTROLAnalysis: Solving Reduced Space Unconstrained Optimization Problem" << std::endl;
+      auto algo = ROL::TypeU::AlgorithmFactory<double>(rolParams.sublist("ROL Options"));
+      algo->run(rol_p_primal, reduced_obj, *rolOutput);
+      return_status = algo->getState()->statusFlag;
+    }
+*/
+    //! check correctness of Gradient prvided by Model Evaluator
+    if(rolParams.get<bool>("Check Derivatives", false)) {
```
Comment (mperego) [inline packages/piro/src/Piro_PerformAnalysis.cpp:1005]: It seems we are only checking the derivatives of the reduced objective. Is it possible to check also the partial derivatives?

File: packages/piro/test/_input_Analysis_ROL_ReducedSpace_Transient_2_parameters.xml [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+          <Parameter name="Test Type" type="string" value="NormF" />
+          <Parameter name="Tolerance" type="double" value="1.0e-8" />
+        </ParameterList>
+        <ParameterList name="Test 1">
+          <Parameter name="Test Type" type="string" value="MaxIters" />
+          <Parameter name="Maximum Iterations" type="int" value="10" />
+        </ParameterList>
+      </ParameterList>
+    </ParameterList>
+    <ParameterList name="Rythmos">
```
Comment (mperego) [inline packages/piro/test/_input_Analysis_ROL_ReducedSpace_Transient_2_parameters.xml:149]: We should remove any sections related to Rhytmos in all the tests

File: packages/piro/test/_input_Analysis_ROL_ReducedSpace_Transient_2_parameters.xml [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+        <Parameter name="Check Derivatives" type="bool" value="true" />
+
+        <Parameter name="Test Vector" type="bool" value="false" />
+        <Parameter name="Number Of Vector Tests" type="int" value="1" />
+
+        <Parameter name="Bound Constrained" type="bool" value="false" />
+
+        <Parameter name="Full Space" type="bool" value="false" />
+        <Parameter name="Use NOX Solver" type="bool" value="false" />
```
Comment (mperego) [inline packages/piro/test/_input_Analysis_ROL_ReducedSpace_Transient_2_parameters.xml:192]: Is this option valid? It seems you use NOX through tempus.

**Review by mperego — COMMENTED** [review]
thanks @kliegeois for putting this together. It's exciting to have this capability.

Besides the inline comments, I have a couple of comments.
I don't understand how the tests work, they seem steady problems to me.
Also might be good to add more testing coverage. It seems we do not have a test for bound constrained inversion.

**Comment by mperego** [comment]
@dridzal , @dpkouri can you have a look at the changes in ROL?

**Review by mperego — COMMENTED** [review]
Looks good. I left some minor comments. You could address then in a following PR.

**Comment by mperego** [comment]
@dpkouri @dridzal can you have a look at the file added to ROL repo?

**Review by dridzal — APPROVED** [review]
This is a reasonable addition.

**Pull request merged**